contents/docs/libraries/js/index.mdx

```diff
@@ -456,6 +456,153 @@ const handleCookieConsent = (consent) => {
 
 - If you don't want PostHog to store anything on the user's browser (e.g. if you want to rely on your own identification mechanism only or want completely anonymous users), you can set `disable_persistence: true` in PostHog's config. If you do this, remember to call [`posthog.identify`](#identifying-users) **every time** your app loads. If you don't, every page refresh is treated as a new and different user.
 
+## Amending events before they are captured
+
+When initializing the SDK you can provide a `beforeSend` function. 
+This can be used to amend or reject events before they are sent to PostHog.
+
+> **Note:** amending and rejecting events should be done with care. It can cause unexpected results in parts of PostHog.
+
+### Not all events are processed by `beforeSend`.
+
+Events which are to do with billing or where sampling/editing would break the PostHog feature are never processed by `beforeSend`
+
+* $web_experiment_applied
+* $$client_ingestion_warning
+* $feature_enrollment_update
+* $feature_flag_called
+* survey dismissed
+* survey sent
+* survey shown
+* $snapshot
+* $identify
+* $groupidentify
+* $create_alias
+
+### Redacting information in events
+
+`beforeSend` gives you one place to edit or redact information before it is sent to PostHog.
+
+#### Redact URLs in event properties
+
+```
+posthog.init('my_token', {
+    beforeSend: (event): event | null => {
+        const redactedProperties = {}
+        for (const [key, value] of Object.entries(event.properties || {})) {
+            if (key.includes(url)) {
+                redactedProperties[key] = value.replace(/customer_id/\d+/, 'customer_id/1234567')
+            } else {
+                redactedProperties[key] = value
+            }
+        }
+        event.properties = redactedProperties
+        
+        return event
+    }
+})
+```
+
+#### Redact person properties in $set or $set_once
+
+```
+posthog.init('my_token', {
+    beforeSend: (event): event | null => {
+        event.$set = {
+            ...event.$set,
+            homeAddress: '******'
+        }
+        event.$set_once = {
```

Comment:
isn't `event.properties.x` always?

Reply:
person updates are carried on `$set` or `$set_once` at the top level of things sent to capture
they don't all have person processing e.g. $snapshot ignore them
bit for some redaction you have to care about things outside of properties

it's why there's a TODO in there - i've never really touched the person stuff and need to confirm how it works